### PR TITLE
Fix ReferenceCounter to support arrays that never be pushed

### PR DESCRIPTION
**Pull request by erikzhang**

### CODE REVIEW

**Review by shargon — COMMENTED** [review]
It's not better to remove `ReferenceCounter` if never will be pushed?

File: tests/neo-vm.Tests/UtReferenceCounter.cs [hunk 1]
```diff
@@ -146,5 +147,19 @@ public void TestRemoveReferrer()
...
+        [TestMethod]
+        public void TestArrayNoPush()
+        {
+            using ScriptBuilder sb = new ScriptBuilder();
+            sb.Emit(OpCode.RET);
+            using ExecutionEngine engine = new ExecutionEngine();
+            engine.LoadScript(sb.ToArray());
+            Assert.AreEqual(0, engine.ReferenceCounter.Count);
+            Array array = new Array(engine.ReferenceCounter, new StackItem[] { 1, 2, 3, 4 });
+            Assert.AreEqual(array.Count, engine.ReferenceCounter.Count);
```
Comment (shargon) [inline tests/neo-vm.Tests/UtReferenceCounter.cs:160]: 4 items + array should be 5, isn't it?

Comment (erikzhang) [reply tests/neo-vm.Tests/UtReferenceCounter.cs:160]: The array is not pushed onto the stack.

**Comment by erikzhang** [comment]
> It's not better to remove ReferenceCounter if never will be pushed?

But you cannot predict whether it will be pushed onto the stack, if this array will be passed as a parameter.

**Comment by erikzhang** [comment]
Go?

**Review by shargon — APPROVED** [review]

**Pull request merged**